arch/arm/mcount-support.c

```diff
@@ -296,7 +296,7 @@ int check_float_abi_cb(struct dl_phdr_info *info, size_t size, void *data)
 	for (i = 0; i < info->dlpi_phnum; i++) {
 		const Elf32_Phdr *phdr = info->dlpi_phdr + i;
 
-		if (phdr->p_type == PT_LOAD) {
+		if (phdr->p_type == PT_LOAD && phdr->p_vaddr) {
 			Elf32_Ehdr *ehdr = (void *)phdr->p_vaddr;
```

Comment:
Hmm.. I guess the check should remain as is but the ehdr pointer should be changed like below:

```c
if (phdr->p_type == PT_LOAD) {
    Elf32_Ehdr *ehdr = (void *)info->dlpi_addr + phdr->p_vaddr;
```

Reply:
Hmm.. I need to test it again. Could you explain about this new change?

Reply:
The dlpi_addr contains address of the module actually loaded.  I guess it was usually 0 for normal executables.

Reply:
I mean that the segfault is occurred at `ehdr->e_flags` in line number 10.
```c
     1  int check_float_abi_cb(struct dl_phdr_info *info, size_t size, void *data)
     2  {
     3          unsigned i;
     4
     5          for (i = 0; i < info->dlpi_phnum; i++) {
     6                  const Elf32_Phdr *phdr = info->dlpi_phdr + i;
     7
     8                  if (phdr->p_type == PT_LOAD) {
     9                          Elf32_Ehdr *ehdr = (void *)phdr->p_vaddr;                                                                                                                                                                                         
    10                          use_hard_float = ehdr->e_flags & EF_ARM_ABI_FLOAT_HARD;
    11                          break;
    12                  }
    13          }
    14          float_abi_checked = true;
    15
    16          return 1;
    17  }
```
> The dlpi_addr contains address of the module actually loaded. I guess it was usually 0 for normal executables.

If both `info->dlpi_addr` and `phdr->p_vaddr` are zero, then `ehdr->e_flags` gets crashed.
Please correct me if I'm wrong.

Reply:
I don't think the both can be zero at the same time.  Did you see the case?

Reply:
Yes, I saw it was crashed so added this check routine.

Reply:
Do you mean `info->dlpi_addr + phdr->p_vaddr == 0`?

Reply:
Yes. It was zero so crashed.

Reply:
Here is the gdb log for this crash.
```
292│ int check_float_abi_cb(struct dl_phdr_info *info, size_t size, void *data)
293│ {
294│         unsigned i;
295│
296│         for (i = 0; i < info->dlpi_phnum; i++) {
297│                 const Elf32_Phdr *phdr = info->dlpi_phdr + i;
298│
299│                 if (phdr->p_type == PT_LOAD) {
300│                         Elf32_Ehdr *ehdr = (void *)phdr->p_vaddr;
301├───────────────────────> use_hard_float = ehdr->e_flags & EF_ARM_ABI_FLOAT_HARD;
302│                         break;
303│                 }
304│         }
305│         float_abi_checked = true;
306│   
307│         return 1;
308│ }
```
```
Thread 2.1 "app_shell" received signal SIGSEGV, Segmentation fault.
[Switching to process 4571]
0x76fc04a8 in check_float_abi_cb (info=<optimized out>, size=<optimized out>, data=<optimized out>) at /usr/src/debug/uftrace/0.9.2-r0/git/arch/arm/mcount-support.c:301
(gdb) bt
#0  0x76fc04a8 in check_float_abi_cb (info=<optimized out>, size=<optimized out>, data=<optimized out>) at /usr/src/debug/uftrace/0.9.2-r0/git/arch/arm/mcount-support.c:301
#1  0x7297339c in dl_iterate_phdr () from /lib/libc.so.6
#2  0x54d53000 in ?? ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

(gdb) p ehdr
$1 = (Elf32_Ehdr *) 0x0
```

Reply:
But you only used `phdr->p_vaddr` here..

Reply:
Actually, I don't see this crash after a few more commits added, but saw it before. So I couldn't test `info->dlpi_addr + phdr->p_vaddr == 0` check.

> The dlpi_addr contains address of the module actually loaded. I guess it was usually 0 for normal executables.

But as you mentioned above, if `dlpi_addr` is normally zero, then there might be some possibility of being crashed because `phdr->p_vaddr` can also be zero. If you're not sure about it, I will drop this, then add this change if I see the same problem next time.

Reply:
No, I think they should not be zero at the same time and it's the right solution IMO.  For PIE `phdr->p_vaddr` could be zero but then `info->dlpi_addr` has the actual load address.  For non-PIE it'll go the other way around.

Reply:
Thank you for the explanation. Then I will drop this although I'm just wondering why it was crashed as shown in the gdb report. I will think about it again when I see the same problem next time. Please never mind this time. Thanks.

Reply:
No no no... please change it to use `info->dlpi_addr + phdr->p_vaddr`.